src/objective/aft_obj.cc

```diff
@@ -0,0 +1,116 @@
+/*!
+ * Copyright 2015 by Contributors
+ * \file rank.cc
+ * \brief Definition of aft loss.
+ */
+
+#include <dmlc/omp.h>
+#include <xgboost/logging.h>
+#include <xgboost/objective.h>
+#include <vector>
+#include <limits>
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <cmath>
+
+#include "xgboost/json.h"
+
+#include "../common/math.h"
+#include "../common/random.h"
+#include "../common/survival_util.h"
+
+using AFTParam = xgboost::common::AFTParam;
+using AFTLoss = xgboost::common::AFTLoss;
+
+namespace xgboost {
+namespace obj {
+
+DMLC_REGISTRY_FILE_TAG(aft_obj);
+
+class AFTObj : public ObjFunction {
+ public:
+  void Configure(const std::vector<std::pair<std::string, std::string> >& args) override {
+    param_.UpdateAllowUnknown(args);
+    loss_.reset(new AFTLoss(param_.aft_noise_distribution));
+  }
+
+  void GetGradient(const HostDeviceVector<bst_float>& preds,
+                   const MetaInfo& info,
+                   int iter,
+                   HostDeviceVector<GradientPair>* out_gpair) override {
+    /* Boilerplate */
+    CHECK_EQ(preds.Size(), info.labels_lower_bound_.Size());
+    CHECK_EQ(preds.Size(), info.labels_upper_bound_.Size());
+
+    const auto& yhat = preds.HostVector();
+    const auto& y_lower = info.labels_lower_bound_.HostVector();
+    const auto& y_higher = info.labels_upper_bound_.HostVector();
+    const auto& weights = info.weights_.HostVector();
+    const bool is_null_weight = weights.empty();
+
+    out_gpair->Resize(yhat.size());
+    std::vector<GradientPair>& gpair = out_gpair->HostVector();
+    CHECK_LE(yhat.size(), static_cast<size_t>(std::numeric_limits<omp_ulong>::max()))
+      << "yhat is too big";
+    const omp_ulong nsize = static_cast<omp_ulong>(yhat.size());
+    double first_order_grad;
```

Comment:
Is this shared by multiple omp threads?

Reply:
Good catch. Let me adopt @RAMitchell's suggestion and define variables in the scope they are used.

Reply:
Thanks.  Maybe some tests defining the expected result are needed.

Reply:
I wrote a bunch of C++ tests to define expected results. It was a super useful exercise, since I found two bugs in the process :)